docassemble/AssemblyLine/custom_jinja_filters.py

```diff
@@ -200,11 +203,140 @@ def catchall_subquestion(value: Any, subquestion: str) -> DACatchAll:
+    the actual value when the document is being generated for final output.
+
+    E.g., to show a placeholder for a signature field when the document is being
+    shown to the signer, but show the actual signature when the document is finalized.
+
+    The default placeholder is "[ signature ]" if the variable name follows the pattern "users[0].signature",
+    or [ variable_name ] if it is not an attribute of an item.
+
+    `i` will be the value from the template's context unless it is explicitly passed,
+    as in an ALDocument's "preview" or "final" values.
+
+    Example:
+        Contents of test_if_final.docx:
+        ```jinja
+        {{ users[0].signature | if_final }}
+        ```
+
+        Returns "[ signature ]" if `i` (passed to the context of the attachment block) is not "final",
+        otherwise the actual value of `users[0].signature`.
+
+        ```yaml
+        ---
+        include:
+        - assembly_line.yml
+        ---
+        mandatory: True
+        code: |
+            preview_screen
+            final_screen
+        ---
+        question: |
+            Here is what it looks like unsigned
+        subquestion: |
+            ${ test_if_final_attachment.as_pdf(key="preview") }
+        continue button field: preview_screen
+        ---
+        question: |
+            Here is what it looks like signed
+        subquestion: |
+            ${ test_if_final_attachment.as_pdf(key="final") }
+        event: final_screen
+        ---
+        objects:
+            - test_if_final_attachment: ALDocument.using(title="test_if_final", filename="test_if_final")
+        ---
+        attachment:
+            variable name: test_if_final_attachment[i]
+            docx template file: test_if_final.docx
+        ```
+
+    Args:
+        context (Jinja2Context): The Jinja2 context, automatically passed by the `pass_context` decorator.
+        value (Any): The original value as passed to the filter.
+        i (str, optional): The current value of `i`. If not provided, it will be fetched from the context.
+        expected_values (Union[str, List[str]], optional): The expected value(s) of `i` to trigger returning `value`.
+            Defaults to "final".
+        placeholder (str, optional): The placeholder string to return if the condition is not met. If
+            not provided, a default placeholder will be generated based on the variable name.
+
+    Returns:
+        Any: The original `value` if `i` matches `expected_values`, otherwise the `placeholder`.
+    """
+    if i is None:
+        i = context.get("i")
+
+    if (
+        i != expected_values
+        if isinstance(expected_values, str)
+        else i not in expected_values
+    ):
+        if isinstance(value, Undefined):
+            if not placeholder:
+                _label = _undefined_label(value)
+                if "." in _label:
+                    _label = _label.split(".")[-1]
```

Comment:
When is there a dot in the label names? The docstring of `_undefined_label` specifically says that there shouldn't be a dot in the undefined name, so this logic being here is kinda confusing.